### PR TITLE
Fix!(parser): attach ORDER, LIMIT, FETCH to set ops instead of rhs query

### PR DESCRIPTION
**Pull request by georgesittas**
Fixes the last item in #2653

### CODE REVIEW

**Comment by georgesittas** [comment]
FYI the reason I included `order` and `fetch` is because Postgres mentions the following in their [docs](https://www.postgresql.org/docs/current/queries-union.html):

> You can also surround an individual query with parentheses. **This is important if the query needs to use any of the clauses discussed in following sections, such as LIMIT**. Without parentheses, you'll get a syntax error, or else the clause will be understood as applying to the output of the set operation rather than one of its inputs.

The "following sections" seem to only contain these three args, as far as "query modifiers" go:

Screenshot 2023-12-12 at 4 45 15 PM

**Comment by georgesittas** [comment]
It seems that most dialects follow Postgres' semantics, i.e. they attach `LIMIT` to the union instead of its rhs. One exception is ClickHouse, which seems to attach it to the rhs in both of the following examples:

```sql
with t1 as (select 1 as x union all select 2), t2 as (select 1 as x) select x from t1 union all select x from t2 limit 1;
with t1 as (select 1 as x union all select 2), t2 as (select 1 as x) select x from t1 union all (select x from t2 limit 1);
```

Another exception is T-SQL, because it uses `TOP` instead of `LIMIT`, so you can't have ambiguity as to where it should be attached. We previously arbitrarily attached these modifiers to the rhs and so transpiling towards T-SQL would basically limit it using `TOP`, but now it's dropped. This means that queries that contain `LIMIT` and set ops will have their rightmost `LIMIT` removed when targeting T-SQL.

If needed, I can add a transformation in T-SQL's generation which converts `X UNION Y LIMIT Z` into `SELECT TOP Z * FROM (X UNION Y)`. I'm currently working on adding a flag to handle the ClickHouse case as well (i.e. keep previous behavior for it).

File: sqlglot/parser.py [hunk 1]
```diff
@@ -3259,21 +3263,34 @@ def _parse_set_operations(self, this: t.Optional[exp.Expression]) -> t.Optional[
...
+        comments = self._prev.comments
+        distinct = self._match(TokenType.DISTINCT) or not self._match(TokenType.ALL)
+        by_name = self._match_text_seq("BY", "NAME")
+        expression = self._parse_set_operations(
+            self._parse_select(nested=True, parse_set_operation=False)
+        )
+
+        set_operation_modifiers: t.Dict[str, t.Any] = {}
+        if self.MODIFIERS_ATTACHED_TO_UNION and isinstance(expression, exp.Expression):
+            for arg in self.UNION_MODIFIERS:
```
Comment (tobymao) [inline sqlglot/parser.py:3281]: can we not do a for loop?

Comment (georgesittas) [reply sqlglot/parser.py:3281]: will be a bit verbose / duplicative because we have to check if the args exist etc

Comment (georgesittas) [reply sqlglot/parser.py:3281]: so the for loop is just to compact that a bit

**Review by tobymao — APPROVED** [review]

**Pull request merged**